src/preprocess/mod.rs

```diff
@@ -1 +1,17 @@
-pub mod links;
+pub use self::links::LinkPreprocessor;
+
+mod links;
+
+use book::Book;
+use errors::*;
+
+use std::path::PathBuf;
+
+pub struct PreprocessorContext {
+    pub src_dir: PathBuf
```

Comment:
It's probably better for `PreprocessorContext` to contain a reference (or clone, if you don't want to deal with lifetimes) to the `Config` and the book's root. That way a preprocessor can hook into `book.toml` for configuration, and the `src_dir` can be discovered by combining `config.book.src` and the book root.

So something like this:

```rust
pub struct PreprocessorContext {
  pub config: Config,
  pub root: PathBuf,
  // maybe also include the `Book` here... I dunno
}
```

Reply:
That makes sense, given the other fields of the `MDBook` shouldn't be needed by the preprocessor. I've made it a clone for now since a borrow would be problematic here:

https://github.com/rust-lang-nursery/mdBook/blob/bf093e2f5f0e8c7f455b6e021cb72adb974836d4/src/book/mod.rs#L205

since we'd have a reference to an inner field and additionally trying to take a reference to itself.